### PR TITLE
Update input plugins default active state & joint handling

### PR DESCRIPTION
**Pull request by jherico**
## Changes

* Remove the concept of 'joints captured'
 * The first valid pose encountered from an input plugin will be used
 * Input plugins should only populate the pose map with valid non-identity transforms
* Migrate spacemouse to input plugin architecture, external project
* Migrate input device menu to `Developer` -> `Avatar`
* Default all input plugins to active

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 12e9fb53f91d5dba58aa171a601dc7c8729f0880 available at: https://gist.github.com/8626889b2c04181775024c6216c52cd4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e2771ad11504cc9a32f4eed2ed6746138ccdcd95 available at: https://gist.github.com/8626889b2c04181775024c6216c52cd4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 4f6d4fc625e39049271ebf5024c929c68ae16f90 available at: https://gist.github.com/8626889b2c04181775024c6216c52cd4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: eac6a62b8a1cb710a726d684b71640d401962752 available at: https://gist.github.com/8626889b2c04181775024c6216c52cd4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 26a34d8c589ef1dea01c766e3f79d4955e88f254 available at: https://gist.github.com/8626889b2c04181775024c6216c52cd4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.dmg

**Comment by ZappoMan** [comment]
@jherico - this needs a test plan

**Comment by jherico** [comment]
## Testing

* Basic functionality should continue to work
* If you wipe your INI file and have Vive controllers or Hydra controllers, they should be active by default.
* Having one Hydra controller uncradled and one Vive controller (for the opposite hand) on should function, and both hands should track the respective controllers.
* If both hydras are uncradled  and both vive controllers are on, whichever route is handled first will take precedence, which currently may be random or deterministic, but regardless is undefined.  However, it should be consistent through the life of the application as long as both plugins are enabled.

**Comment by hifi-gustavo** [comment]
Build console output for commit: c9d65c74064514a1b6fa15cb805f46687f06a7b7 available at: https://gist.github.com/8626889b2c04181775024c6216c52cd4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.dmg

File: plugins/openvr/src/ViveControllerManager.cpp [hunk 1]
```diff
@@ -214,12 +212,13 @@ void ViveControllerManager::renderHand(const controller::Pose& pose, gpu::Batch&
 }
 
 
-void ViveControllerManager::pluginUpdate(float deltaTime, const controller::InputCalibrationData& inputCalibrationData, bool jointsCaptured) {
+void ViveControllerManager::pluginUpdate(float deltaTime, const controller::InputCalibrationData& inputCalibrationData) {
+    _inputDevice->update(deltaTime, inputCalibrationData);
```
Comment (Atlante45) [inline plugins/openvr/src/ViveControllerManager.cpp]: Is that update necessary?

File: plugins/hifiSpacemouse/src/SpacemouseManager.cpp [hunk 2]
```diff
@@ -111,70 +163,79 @@ controller::Input::NamedPair SpacemouseDevice::makePair(SpacemouseDevice::Positi
...
 }
 
-void SpacemouseManager::ManagerFocusOutEvent() {
-    instance->focusOutEvent();
-}
+#ifdef Q_OS_WIN
 
-void SpacemouseManager::init() {
-}
+#include <VersionHelpers.h>
```
Comment (Atlante45) [inline plugins/hifiSpacemouse/src/SpacemouseManager.cpp]: Include should be at the top of the file.

File: plugins/hifiNeuron/src/NeuronPlugin.cpp [hunk 3]
```diff
@@ -548,16 +453,23 @@ QString NeuronPlugin::InputDevice::getDefaultMappingConfig() const {
...
 void NeuronPlugin::InputDevice::update(float deltaTime, const controller::InputCalibrationData& inputCalibrationData, const std::vector<NeuronPlugin::NeuronJoint>& joints, const std::vector<NeuronPlugin::NeuronJoint>& prevJoints) {
     for (size_t i = 0; i < joints.size(); i++) {
+        int poseIndex = neuronJointIndexToPoseIndex((NeuronJointIndex)i);
         glm::vec3 linearVel, angularVel;
-        glm::vec3 pos = joints[i].pos;
-        glm::quat rot = eulerToQuat(joints[i].euler);
+        const glm::vec3& pos = joints[i].pos;
+        const glm::vec3& rotEuler = joints[i].euler;
+
+        if ((Vectors::ZERO == pos && Vectors::ZERO == rotEuler)) {
```
Comment (Atlante45) [inline plugins/hifiNeuron/src/NeuronPlugin.cpp]: Are we allowing this now? I vaguely remember a debate about that, can't remember if it was accepted or not.
Extra parenthesis.

File: plugins/openvr/src/ViveControllerManager.cpp (hunk 1 above)
Comment (jherico) [inline plugins/openvr/src/ViveControllerManager.cpp]: looks like a merge conflict resolution failure.  will fix.

File: plugins/hifiSpacemouse/src/SpacemouseManager.cpp (hunk 2 above)
Comment (jherico) [inline plugins/hifiSpacemouse/src/SpacemouseManager.cpp]: fixed

File: plugins/hifiNeuron/src/NeuronPlugin.cpp (hunk 3 above)
Comment (jherico) [inline plugins/hifiNeuron/src/NeuronPlugin.cpp]: fixed the parens.  Not sure what you mean by the other thing... allowing *what*?

**Comment by hifi-gustavo** [comment]
Build console output for commit: 3586991d016d8654306733cc4d3ae7040f37bab4 available at: https://gist.github.com/8626889b2c04181775024c6216c52cd4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: cfb6f708a17548065ddaac454bcbe07fd60c7f71 available at: https://gist.github.com/8626889b2c04181775024c6216c52cd4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.dmg

File: plugins/hifiNeuron/src/NeuronPlugin.cpp (hunk 3 above)
Comment (Atlante45) [inline plugins/hifiNeuron/src/NeuronPlugin.cpp]: Meant this: https://github.com/highfidelity/hifi/wiki/Coding-Standard#343-write-the-expression-of-a-conditional-similar-to-how-you-would-speak-it-out-loud

We debated that rule a couple times, not sure what we landed on.

**Comment by ZappoMan** [comment]
I'd really like to get this into the next RC release.

**Comment by Atlante45** [comment]
@ZappoMan Don't have a spacemouse, but I can test the Vive/Hydra related items.

**Comment by kevinmthomas-carpool** [comment]
FAIL.  Unable to complete test as the mouse is completely unusable as Hydras hijack Interface.

**Comment by jherico** [comment]
@kevinmthomas-carpool can you retest with the latest once it builds?

**Comment by Atlante45** [comment]
@jherico Was able to switch between Hydras/Vive controller no problem.
Tried the 1 vive/1 hydra think and I couldn't get it to work, also I wonder why anybody would want to do that xD. Basically only one would ever work at the time tracking wise. Buttons would work for both at the same time.

A small oddity I noticed with vive controllers. If shut down my left controller, the right controller tracks the right hand no problem. But if I shut down my right controller, then the left controller tracks both hands at the same time.

**Comment by kevinmthomas-carpool** [comment]
@jherico will do!

**Comment by hifi-gustavo** [comment]
Build console output for commit: 73342b2758c1ae89188186a1b23b5e2b8b16664e available at: https://gist.github.com/8626889b2c04181775024c6216c52cd4<br/><br/>Installers available for download at: <br/>Windows: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.exe<br/>OS X: http://cachefly.highfidelity.com/HighFidelity-Beta-PR7870.dmg

**Comment by Atlante45** [comment]
Tested the space mouse and I can only move my camera back and forth from 1st to 3rd person with it.

**Comment by kevinmthomas-carpool** [comment]
## QA Test
**QA Tester**: KevinMThomas
**localhost 'createTestBlocks.js' Result**: **localhost 'createTestBlocks.js'** SUCCESS
**localhost ATP Upload + Edit Result**: **localhost ATP** SUCCESS
**Echo Server Audio**: **Echo Server Audio** SUCCESS
**localhost 'harmonicOscillator.js' Result**: **localhost 'harmonicOscillator.js'** SUCCESS
**localhost 'reverbTest.js' Result**: **localhost 'reverbTest.js'** SUCCESS
**localhost 'playaPerformanceTest.js' Result**: **localhost 'playaPerformanceTest.js'** SUCCESS
**playa 'playaPerformanceTest.js' Result**: **playa 'playaPerformanceTest.js'** SUCCESS
**cellscience 'playaPerformanceTest.js' Result**: **cellscience 'playaPerformanceTest.js'** SUCCESS
**QA Pass**: **QA Pass** SUCCESS
**Additional Test Results + Observations**: Brilliant improvement on interaction of hydras and mouse. Was able to put hydras on floor and out of range and mouse worked perfectly! Great enhancement and integration of hand controllers and mouse.

**Domain** |localhost |Playa |cellscience
---|---|---|---
**Render Rate**: |60 |61 |53
**Present Rate**: |60 |60 |60
**Present New Rate**: |60 |54 |53
**CPU Buffer Count**: |4227 |1748 |1861
**GPU Buffer Count**: |1080 |301 |255
**CPU Memory Usage**: |48553984 |40574976 |43405312
**GPU Memory Usage**: |27172864 |14651392 |9146368
**CPU Texture Count**: |249 |115 |90
**GPU Texture Count**: |713 |609 |593
**GPU Transfer Count**: |0 |0 |0
**CPU Texture Memory Usage**: |122955772 |239033664 |55716936
**GPU Texture Memory Usage**: |755525990 |162876570 |60645468
**GPU Virtual Memory Usage**: |893213030|300563610|198332508
**Triangle Count**: |559179 |335865 |273411
**Triangle Rate**: |34948687 |20991562 |16083000
**API Drawcall Count**: |259 |77 |51
**Drawcall Count**: |263 |176 |51
**Drawcall Rate**: |16437 |11000 |3000
**State Changes Texture Count**:|259 |81 |40
**Pipeline Count**: |30 |23 |16
**Item Opaques**: |317 |137 |50
**Item Translucents**: |14 |5 |14
**Item Lights**: |25 |21 |0
**Timing Opaques**: |2062 |1066 |551
**Timing Translucents**: |169 |80 |165
**Timing Lighting**: |152 |184 |40
**Render Frame**: |3348 |2592 |1575

**Comment by Nex-Pro** [comment]
@Atlante45 I have a Space Navigator it uses the spacemouse drivers.

**Comment by Atlante45** [comment]
@Nex-Pro Yeah, I had to install the driver to get it to work. But it appears the space mouse is broken in the current release built. Filed a bug about it.

**Pull request merged**